Firestore/core/src/firebase/firestore/util/log_apple.mm

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2017 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+
+#include <string>
+
+#import <FirebaseCore/FIRLogger.h>
+
+#import "Firestore/Source/API/FIRFirestore+Internal.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+void LogSetLevel(LogLevel level) {
+  switch (level) {
+    case kLogLevelVerbose:
+      FIRSetLoggerLevel(FIRLoggerLevelMax);
+      break;
+    case kLogLevelDebug:
+      FIRSetLoggerLevel(FIRLoggerLevelDebug);
+      break;
+    case kLogLevelInfo:
+      FIRSetLoggerLevel(FIRLoggerLevelInfo);
+      break;
+    case kLogLevelWarning:
+      FIRSetLoggerLevel(FIRLoggerLevelWarning);
+      break;
+    case kLogLevelError:
+      FIRSetLoggerLevel(FIRLoggerLevelError);
+      break;
+    case kLogLevelAssert:
+      // Unsupported log level. FIRSetLoggerLevel will deal with it.
+      FIRSetLoggerLevel(-1);
+      break;
+    default:
+      // Unsupported log level. FIRSetLoggerLevel will deal with it.
+      FIRSetLoggerLevel(-1);
+      break;
+  }
+}
+
+LogLevel LogGetLevel() {
+  // We return the true log level. True log level is what the SDK used to
+  // determine whether to log instead of what parameter is used in the last call
+  // of LogSetLevel().
+  if (FIRIsLoggableLevel(kLogLevelInfo, NO)) {
+    if (FIRIsLoggableLevel(FIRLoggerLevelDebug, NO)) {
+      // FIRLoggerLevelMax is actually kLogLevelDebug right now. We do not check
+      // further.
+      return kLogLevelDebug;
+    } else {
+      return kLogLevelInfo;
+    }
+  } else {
+    if (FIRIsLoggableLevel(kLogLevelWarning, NO)) {
+      return kLogLevelWarning;
+    } else {
+      return kLogLevelError;
+    }
+  }
+}
+
+void LogDebug(const char* format, ...) {
+  va_list list;
+  va_start(list, format);
+  FIRLogBasic(FIRLoggerLevelDebug, kFIRLoggerFirestore, @"I-FST000001",
+              @(format), list);
```

Comment:
Using an Objective-C boxed expression in `@(format)` causes the allocated NSString to copy the contents of the format. You can avoid the copy by defining a function like this:

```
static NSString* FormatString(const char* format) {
  return [[NSString alloc] initWithBytesNoCopy:format
                                        length:strlen(format)
                                      encoding:NSUTF8StringEncoding
                                  freeWhenDone:NO];
}
```